scripts/buildorder.py

```diff
@@ -267,7 +268,10 @@ def generate_full_buildorder(pkgs_map):
     build_order = []
 
     # List of all TermuxPackages without dependencies
-    leaf_pkgs = [pkg for pkg in pkgs_map.values() if not pkg.deps]
+    leaf_pkgs = []
+    for pkg in pkgs_map.values():
+        if pkg not in leaf_pkgs:
+            leaf_pkgs.append(pkg)
```

Comment:
AIUI, the current behavior is:

1) Find all packages with no dependencies. This is the initial queue.
2) While the queue isn't empty, pop package `pkg` and:
  a) Append `pkg` to the build order.
  b) Remove `pkg` from the un-build-orefered dependencies of all its revdeps, which we track in a map.
  c) If any of the revdeps now have all their deps in the build order, add them to the queue.

The "cycle detection" isn't exactly that - what it actually detects is that process failing to put all the packages in `pkgs_map` into the build queue. A real cycle bug would cause that, but my understanding in #20338 is that it happens because, when you run `build-all.sh`, you run in a mode where subpkgs get installed into `pkgs_map` as aliases for their superpkgs (i.e., `pkgs_map["vim-runtime"] = TermuxPackage("vim")`). The `vim` buildspec really does depend on a pkg called `vim-runtime`, but because it's installed as an alias, the subpkg `vim-runtime` never gets picked up by the algorithm I described. That's the problem, and I added the cycle printer because that's what I needed to figure that out.

So:
1) If you are solving that problem, the commit message (preferable imo) or PR should say how you're solving it. Does the code now notice subpkgs that are leaves and add them to the initial build order? Are you adding special handling for packages that have no deps left un-build-ordered except for subpkgs? This shouldn't be a collection of line-level changes, this should be a higher-level description of the behavior change, similar to what I wrote above.

2) If you have solved it in one of those ways, something's wrong in this block. It no longer does what its comment says (it appears that `leaf_pkgs` now contains all packages, regardless of whether they're leaves), and the thing the comment says it does is a critical step of how this function works. You'll need make sure the code around the code you changed reflects your changes.

Reply:
After reading your long, helpful comment, I realized that the `generate_target_buildorder` function requires rewriting the logic.